File: cpp/649.dota2-senate.cpp
class Solution {
public:
    string predictPartyVictory(string senate) {
        int r = 0, d = 0,
            ban_r = 0, ban_d = 0;

        
        queue<char> que;

        for (char sen : senate) {
            que.push(sen);
            if (sen == 'R') {
                r++;
            } else {
                d++;
            }
        }

        while (!senate.empty()) {
            char sen = que.front(); que.pop();

            if (sen == 'R' && r > 0) {
                if (ban_r > 0) {
                    ban_r--;
                    continue;
                }

                if (d > 0) {
                    d--;
                    ban_d++;
                    que.push(sen);
                    continue;
                }

                return string("Radiant");

            } else if (sen == 'D' && d > 0) {
                if (ban_d > 0) {
                    ban_d--;
                    continue;
                }

                if (r > 0) {
                    r--;
                    ban_r++;
                    que.push(sen);
                    continue;
                }
                return string("Dire");
            }
        }

        return d > 0 && ban_d == 0 ? string("DDD") : string("RRR");
    }
};

File: python/649.dota2-senate.py
class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        r, d = 0, 0
        for senator in senate:
            if senator == 'R':
                r += 1
            else:
                d += 1

        ban_r, ban_d = 0, 0

        senate = deque(senate)
        while senate:
            sen = senate.popleft()

            if sen == 'R' and r > 0:
                if ban_r > 0:
                    ban_r -= 1
                    continue

                if d > 0:
                    d -= 1
                    ban_d += 1
                    senate.append(sen)
                    continue

                return 'Radiant'

            elif sen == 'D' and d > 0:
                if ban_d > 0:
                    ban_d -= 1
                    continue
                
                if r > 0:
                    r -= 1
                    ban_r += 1
                    senate.append(sen)
                    continue

                return 'Dire'

            # return 'Radiant' if d == 0 else 'Dire'